vimeo/VimeoUpload
# VIM-6117: Deprecate upload ticket

#### Ticket

[VIM-6117](https://vimean.atlassian.net/browse/VIM-6117)

#### Pull Request Checklist


#### Issue Summary

We need to upgrade to API version 3.4 for the total storage quota feature. One change that comes with this upgrade is that we no longer receive an upload ticket in response to creating a video. We now receive a video. On this video is an "upload" object that encompasses much of the data formerly help on the upload ticket.

This ticket implements the necessary changes for receiving a video object instead of the upload ticket upon video creation. Because old upload does not support this, we also need to freeze its API version at < 3.4 (we just keep it at 3.2 for now). 

#### Implementation Summary

- Because we now receive a video instead of an upload ticket, the type of object stored in memory was changed to `VIMVideo` for the following classes: `VideoSettingsViewController`, `UploadDescriptor`, `CreateVideoOperation`, `ExportSessionExportCreateVideoOperation.swift`
- `VimeoSessionManager+Upload.swift` - now handling receipt of a video instead of an upload ticket. this only affects new upload. 
- `Podfile` - updated podfile to consume a feature branch of `VimeoNetworking` See changes this branch encompasses here: https://github.com/vimeo/VimeoNetworking/pull/226
This branch adds the new upload object class and also allows us to specify the API version for the request serializer from the creation site of the session manager, giving clients more transparent control over what API version they are consuming. 
- `OldVimeoUploader` & `NewVimeoUploader`: specified the API version that's needed for each singleton
- Migration of old to new objects and vice versa: We persist some objects via our descriptor system. In `UploadDescriptor, we no longer encode a ticket object, but now a video. When decoding we need to convert decoded ticket objects to videos.  In `AppDelegate` & `NSKeyedUnarchiver+Migrations`, we take care of another case, where the uploadQuota was archived as an Objective-C class but unarchived as a Swift class. See inline comment for more info. 

#### Reviewer Tips

- Note this was a big upgrade to `VimeoNetworking`, so many of the file changes in VimeoNetworking reflect that. In fact, the reviewer can ignore all changes to `VimeoNetworking` in this PR. The relevant changes are in the PR linked here: https://github.com/vimeo/VimeoNetworking/pull/226

- In previous commits, I implemented a protocol based approach that defined "the properties required to complete an upload" instead of being tied to a specific object like a video or ticket, but `UploadDescriptor` is ultimately dependent on having a video. So the reactor scope became too large to tackle it in this PR. 

#### How to Test
- Generate a token using your apps page on https://developer.vimeo.com/apps. Make sure it has upload as a scope. Lmk if you need a hand with this. 
- Replace the instances of "YOUR_OAUTH_TOKEN" for old and new upload. 
- Confirm you are able to upload a video using both the old and new projects. 
- Bonus test for migration: build on develop and interrupt an upload by quitting the app. Build on top of that app with this branch, and confirm app does not crash.
